### PR TITLE
fix(perf): remove @st.cache_resource from train_model and add load_pipeline

### LINKED ISSUE #99: [Bug] train_model() sklearn pipeline is trained and cached at every cold start but never used for prediction — wastes 20-30s startup time with dead code
## Description

`application.py` calls `train_model()` at module load time, which reads both CSV files, engineers all features, and fits a full XGBoost pipeline inside a `@st.cache_resource` block:

```python
# application.py line ~1076
@st.cache_resource
def train_model():
    matches = pd.read_csv("matches.csv")
    deliveries = pd.read_csv("deliveries.csv")
    # ... full feature engineering ...
    pipe = Pipeline([
        ('preprocessor', preprocessor),
        ('model', XGBClassifier(n_estimators=100, ...))
    ])
    pipe.fit(X, y)
    return pipe

pipe = train_model()  # Called at startup — takes ~20-30s on cold start
```

However, `pipe` is **never referenced again** in the entire file. When the user clicks **Run Analysis**, the prediction code bypasses `pipe` entirely and loads a completely different model from disk:

```python
# application.py — inside the `if analyze:` block
loaded_xgb = xgb.XGBClassifier()
loaded_xgb.load_model('xgb_win_prob_model.json')  # Loads separate file every click
```

The two models are also incompatible — `train_model()` uses sklearn feature names (`target`, `wickets`) while the prediction manually builds a feature dict with different names (`target_score`, `wickets_left`). Passing `pipe.predict()` the prediction input would raise a feature name mismatch error.

## Impact

1. **Startup performance**: Every Streamlit cold start (server restart, Streamlit Cloud deployment, sleep-wake cycle) triggers a full model train on ~750k delivery rows. This adds 20-30 seconds of blocking startup time that provides zero benefit to the user.
2. **Dead code**: `pipe` is assigned but never read. Any future contributor who attempts to use `pipe` for predictions will get incorrect results due to the feature name mismatch.
3. **Double resource usage**: The app loads both the sklearn pipeline into memory AND the XGBoost JSON model, consuming roughly double the RAM needed.

## Steps to Reproduce

1. Restart the Streamlit server (or clear the cache with `st.cache_resource.clear()`).
2. Open the app.
3. Observe the ~20-30 second startup delay.
4. Search the entire `application.py` file for usages of the variable `pipe` — it is assigned once and never read.

## Proposed Fix

Remove `train_model()` entirely (or keep it only in the notebook for experimentation). At startup, load the pre-trained `xgb_win_prob_model.json` once using `@st.cache_resource`:

```python
@st.cache_resource
def load_model():
    model = xgb.XGBClassifier()
    model.load_model('xgb_win_prob_model.json')
    return model

loaded_xgb = load_model()  # Cached — loaded once, reused on every prediction
```

Then in the prediction block, replace:
```python
# Remove these two lines:
loaded_xgb = xgb.XGBClassifier()
loaded_xgb.load_model('xgb_win_prob_model.json')
```
with the cached `loaded_xgb` from above.

This reduces cold start time from ~30s to under 1 second and eliminates the dead training code.

## Affected File

`application.py` — `train_model()` function (lines ~1076-1125) and the prediction block (lines ~1522-1524)

## Suggested Labels

`type:bug` `level:advanced` `quality:exceptional`

I would like to work on this issue under GSSoC 2026. Could you please assign it to me? @Arnav-Singh-5080

/assign

### PR DESCRIPTION
**Pull request by anshul23102**
## Summary

Fixes #99

### Root cause

`train_model()` was annotated with `@st.cache_resource` and called at module level (`pipe = train_model()`). On every cold start (server restart, Streamlit Cloud wake-up, cache clear), Streamlit executed the full function body: reading `matches.csv` + `deliveries.csv`, merging and engineering features across ~750k delivery rows, and fitting a `LogisticRegression` pipeline. This blocked app startup for a significant time with zero benefit, because a pre-trained `pipe.pkl` is already committed to the repository and `joblib` is listed in `requirements.txt`.

### What was wrong with using `@st.cache_resource` on `train_model()`

`@st.cache_resource` is designed for objects that are expensive to create and should be shared across sessions. Wrapping a full training pipeline in it means the expensive work runs unconditionally on every cold start instead of being bypassed by loading the artifact. The decorator provides session-sharing but does not prevent the initial training cost.

### Changes in `application.py`

**`train_model()` (demoted to offline utility)**
- Removed `@st.cache_resource`
- Renamed internal variable from `pipe` to `trained` to avoid shadowing the module-level name
- Added a comment marking it as an offline tool for regenerating `pipe.pkl`; it is no longer called at app startup

**`load_pipeline()` (new, `@st.cache_resource`)**
- Primary path: `joblib.load("pipe.pkl")` (sub-second deserialisation)
- Fallback: if `pipe.pkl` is absent, calls `train_model()` once and immediately writes the result back to `pipe.pkl` via `joblib.dump()`, so all subsequent cold starts use the fast path automatically
- `@st.cache_resource` pins the object across all sessions and reruns with no repeat disk I/O

**Call site**
- `pipe = train_model()` replaced with `pipe = load_pipeline()`

**`import joblib`** added (already in `requirements.txt`)

### Cold start comparison

| | Before | After |
|---|---|---|
| `pipe.pkl` present | Trains from CSV anyway (cache miss on first start) | Loads artifact in under 1s |
| `pipe.pkl` absent | Trains from CSV | Trains once, saves `pipe.pkl`, then fast on next start |
| Streamlit rerun | Reads from `@st.cache_resource` memory | Reads from `@st.cache_resource` memory |

## Test plan

- [ ] Delete `pipe.pkl`, restart the app: confirms fallback trains and writes `pipe.pkl`
- [ ] Restart again with `pipe.pkl` present: confirms sub-second startup, no CSV reads
- [ ] Run Analysis with valid inputs: confirms `pipe.predict_proba` returns correct probabilities
- [ ] Call `st.cache_resource.clear()` and reload: confirms `load_pipeline()` re-reads `pipe.pkl` cleanly

### CODE REVIEW

**Comment by Arnav-Singh-5080** [comment]
Closing this pr due to merge conflicts
